### PR TITLE
Release notes based on Robo

### PR DESCRIPTION
**Pull request by AronNovak**
#70 

## Status
image

### CODE REVIEW

File: RoboFile.php [hunk 1]
```diff
@@ -753,4 +753,40 @@ protected function getUserPassword($site, $environment) {
...
+   *   The git tag to compare since. Usually the tag from the previous release.
+   *   If you're releasing for example 1.0.2, then you should get changes since
+   *   1.0.1, so $tag = 1.0.1.
+   *
+   * @throws \Exception
+   */
+  public function generateReleaseNotes($tag) {
+    $result = $this->_exec("git tag | grep $tag || exit 1");
+    if ($result->getExitCode()) {
+      throw new Exception('The tag does not exist.');
```
Comment (amitaibu) [inline RoboFile.php:769]: What would be super handy, is if $tag = null, then find the recent tag and ask if we'd want to compare from that one

Comment (AronNovak) [reply RoboFile.php]: image

Comment (amitaibu) [reply RoboFile.php]: Cool. Is there any easy way to avoid the exit 1 which looks like an error?

image

File: RoboFile.php [hunk 2]
```diff
@@ -753,4 +753,67 @@ protected function getUserPassword($site, $environment) {
...
+      }
+
+      // The issue number is a required part of the branch name
+      // So usually we can grab it from the log too, but that's optional
+      // If we cannot detect it, we still print a less verbose changelog line.
+      $issue_matches = [];
+      preg_match_all('!from Gizra/([0-9]+)!', $line, $issue_matches);
+
+      if (isset($issue_matches[1][0])) {
+        print "- Issue #{$issue_matches[1][0]} :{$log_messages[1]} (Pull Request #{$pr_matches[1][0]})\n";
```
Comment (amitaibu) [inline RoboFile.php:811]:
```suggestion
        print "- Issue #{$issue_matches[1][0]} :{$log_messages[1]} (#{$pr_matches[1][0]})\n";
```

File: RoboFile.php [hunk 3]
```diff
@@ -753,4 +753,67 @@ protected function getUserPassword($site, $environment) {
...
+      }
+      else {
+        print "- {$log_messages[1]} (Pull Request #{$pr_matches[1][0]})\n";
```
Comment (amitaibu) [inline RoboFile.php:814]:
```suggestion
        print "- {$log_messages[1]} (#{$pr_matches[1][0]})\n";
```

File: RoboFile.php [hunk 4]
```diff
@@ -753,4 +753,67 @@ protected function getUserPassword($site, $environment) {
...
+    $this->say('Copy release notes below');
+    echo "Changelog:\n";
+
+    foreach ($lines as $line) {
+      $log_messages = explode("¬¬|¬¬", $line);
+      $pr_matches = [];
+      preg_match_all('/Merge pull request #([0-9]+)/', $line, $pr_matches);
+
+      if (count($log_messages) < 2) {
+        continue;
```
Comment (amitaibu) [inline RoboFile.php:792]: Lets have comments above to those `continue`

File: RoboFile.php [hunk 5]
```diff
@@ -753,4 +753,67 @@ protected function getUserPassword($site, $environment) {
     return $credentials[$environment];
   }
 
+  /**
+   * Generates log of changes since the given tag.
+   *
+   * @param string $tag
+   *   The git tag to compare since. Usually the tag from the previous release.
```
Comment (amitaibu) [inline RoboFile.php:760]: Lets update the doc also about the auto-find of last tag

File: RoboFile.php [hunk 6]
```diff
@@ -753,4 +753,67 @@ protected function getUserPassword($site, $environment) {
...
+   * @throws \Exception
+   */
+  public function generateReleaseNotes($tag = NULL) {
+    $result = $this->_exec("git tag | grep $tag || exit 1");
+    if ($result->getExitCode()) {
+      $latest_tag = $this->taskExec("git tag --sort=version:refname | tail -n1")->printOutput(FALSE)->run()->getMessage();
+      if (empty($latest_tag)) {
+        throw new Exception('There are no tags in this repository.');
+      }
+      if (!$this->confirm("Would you like to compare from $latest_tag?")) {
```
Comment (amitaibu) [inline RoboFile.php:773]:
```suggestion
      if (!$this->confirm("Would you like to compare from the latest tag: $latest_tag?")) {
```

File: RoboFile.php [hunk 7]
```diff
@@ -753,4 +753,67 @@ protected function getUserPassword($site, $environment) {
...
+        $this->say("Specify the tag as an argument");
+        exit(1);
+      }
+      $tag = $latest_tag;
+    }
+
+    $log = $this->taskExec("git log --merges --pretty=format:'%s¬¬|¬¬%b' $tag..")->printOutput(FALSE)->run()->getMessage();
```
Comment (amitaibu) [inline RoboFile.php:780]: btw, what's the purpose of `¬¬` here?

Comment (AronNovak) [reply RoboFile.php:780]: It's a supposedly conflict-free delimiter.

File: RoboFile.php (hunk 4 above)
Comment (AronNovak) [reply RoboFile.php:806]: done

File: RoboFile.php (hunk 5 above)
Comment (AronNovak) [reply RoboFile.php:760]: done

File: RoboFile.php (hunk 1 above)
Comment (AronNovak) [reply RoboFile.php]: image

**Comment by amitaibu** [comment]
Thanks. Lets have a follow up to add an entry in the README

**Pull request merged**